### DISCUSSION TITLE
What is external code execution in the `insecure-external-code-execution` option on Dependabot?

### DISCUSSION THREAD

**Discussion by yykamei** [community/community#58657] [Code Security] [answered]
### Select Topic Area

Question

### Body

In some `package-ecosystem`, we have to configure `registries` as well as [`insecure-external-code-execution`](https://docs.github.com/en/code-security/dependabot/dependabot-version-updates/configuration-options-for-the-dependabot.yml-file#insecure-external-code-execution) to `allow` in order to let Dependabot access the private registries. If the two configurations are properly set, the document says like this:

> external code execution that occurs will only have access to the package managers in the registries associated with that `updates` setting.

I'm curious what is external code execution in this explanation. I created a [Ruby Gem](https://github.com/yykamei/t34_gem) that triggers an HTTP request when it's installed (the code is [here](https://github.com/yykamei/t34_gem/blob/26394ebf1aab45d0fe495898b9a80e551bfdcb01/t34_gem.gemspec#L6-L8)), but it looks like not being prevented from accessing an external server. Even though I set `insecure-external-code-execution` to `deny`, Dependabot seemed to provoke an HTTP request; I expected Dependabot would fail. I wonder what is external code execution. What's the purpose of `insecure-external-code-execution`?

Thanks,

**Comment by victorpreston** [comment]
The purpose of the "insecure-external-code-execution" configuration is to control and restrict the execution of external code by dependencies. When this setting is properly configured, it helps ensure that any external code executed during package installation or update is limited to accessing the package managers and registries associated with the updates setting.

**Reply by yykamei** [reply]
Thank you! Do you know what is "external code execution"? I tried [this code](https://github.com/yykamei/t34_gem/blob/26394ebf1aab45d0fe495898b9a80e551bfdcb01/t34_gem.gemspec#L6-L8) in my Ruby Gem to perform "external code execution", and it ran successfully even though `insecure-external-code-execution` is configured as `deny`. Isn't this an "external code execution"?

**Comment by bogn83** [comment]
We are also puzzled about the actual effects and concrete risks to e.g. GITHUB_TOKEN of the repo that dependabot is running for.

Here are the concrete questions:
- Is `GITHUB_TOKEN` of the current repository also part of those stealable credentials OR STRICTLY ONLY what was entered with the Dependabot secrets UI?
- How are the secrets injected into dependabot, can an attacker simply enumerate them or do they need to know the name of the secret?

**Comment by joevin-slq-docto** [comment, marked as answer]
@bogn83 @yykamei you can understand why `unexpected_external_code` can be raised by reading this:
- https://github.com/dependabot/dependabot-core/blob/main/bundler/lib/dependabot/bundler/file_parser.rb#L37-L45
- https://github.com/dependabot/dependabot-core/blob/main/bundler/spec/dependabot/bundler/file_parser_spec.rb#L333-L359

**Reply by yykamei** [reply]
Thank you. This is very helpful.

**Reply by bogn83** [reply]
Thanks for the links. A problem with those links is that they're on main and that the code has changed meanwhile. That's why many people always press y when browsing files and copying the links to them. This way your browser URL will be on the current revision.

I think the links matching the time above are:
https://github.com/dependabot/dependabot-core/blob/5abeda197e44c41c1d7989ab8c13a7c7b36df681/bundler/lib/dependabot/bundler/file_parser.rb#L37-L45
https://github.com/dependabot/dependabot-core/blob/704fa1f5fb24f502b8e8a67bb6214dc1ace39250/bundler/spec/dependabot/bundler/file_parser_spec.rb#L333-L359

**Reply by bogn83** [reply]
Everyone please consider upvoting here https://github.com/orgs/community/discussions/33750 or better yet create a new topic that may be read more likely. Or maybe you have a better idea to get the dependabot team attention.

Allowing insecure code execution is not something that only yykamei and me resorted to because we researched too little or something like that. I've seen plenty of people specifying it with bundler in the dependabot-core repo issues and the docs actually suggest it's what you should do when you have private gem registries and bundler updates.

This whole thing is somewhat of a rabbit hole apparently when looking at [dependabot-core issues](https://github.com/dependabot/dependabot-core/issues?q=is%3Aissue+insecure-external-code-execution) (I'm not indicating accidental complexity).

**Comment by Saidshire21** [comment]
l
